File: gemsfx/src/main/java/com/dlsc/gemsfx/YearView.java
package com.dlsc.gemsfx;

import com.dlsc.gemsfx.skins.YearViewSkin;
import javafx.beans.binding.Bindings;
import javafx.beans.property.*;
import javafx.css.CssMetaData;
import javafx.css.Styleable;
import javafx.css.StyleableIntegerProperty;
import javafx.css.StyleableProperty;
import javafx.css.converter.SizeConverter;
import javafx.scene.control.Control;
import javafx.scene.control.Skin;

import java.time.Year;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/**
 * A view for selecting a year.
 */
public class YearView extends Control {

    /**
     * Constructs a new instance.
     */
    public YearView() {
        getStyleClass().add("year-view");

        setFocusTraversable(false);

        year.bind(Bindings.createIntegerBinding(() -> {
            Year value = getValue();
            if (value != null) {
                return value.getValue();
            }
            return -1;
        }, valueProperty()));
    }

    @Override
    protected Skin<?> createDefaultSkin() {
        return new YearViewSkin(this);
    }

    @Override
    public String getUserAgentStylesheet() {
        return Objects.requireNonNull(YearView.class.getResource("year-view.css")).toExternalForm();
    }

    private final ReadOnlyIntegerWrapper year = new ReadOnlyIntegerWrapper(this, "year");

    public final int getYear() {
        return year.get();
    }

    /**
     * A read-only integer representation of the current value  / year.
     *
     * @return the year as an integer value
     */
    public final ReadOnlyIntegerProperty yearProperty() {
        return year.getReadOnlyProperty();
    }

    private final ObjectProperty<Year> value = new SimpleObjectProperty<>(this, "value", Year.now());

    public final Year getValue() {
        return value.get();
    }

    /**
     * The currently selected value / year.
     *
     * @return the currently selected value / year
     */
    public final ObjectProperty<Year> valueProperty() {
        return value;
    }

    public final void setValue(Year value) {
        this.value.set(value);
    }

    private final StyleableIntegerProperty cols = new StyleableIntegerProperty(4) {
        @Override
        protected void invalidated() {
            if (get() < 1) {
                throw new IllegalArgumentException("number of columns must be larger than 0");
            }
        }
        @Override
        public Object getBean() { return YearView.this; }
        @Override
        public String getName() { return "cols"; }
        @Override
        public CssMetaData<? extends Styleable, Number> getCssMetaData() {
            return StyleableProperties.COLS;
        }
    };

    public final int getCols() {
        return cols.get();
    }

    /**
     * Determines how many columns of years will be displayed. The number of
     * columns multiplied with the number of rows determines the total number
     * of years shown per "page".
     * <p>
     * Can be set via CSS using the {@code -fx-cols} property.
     * Valid values are positive integers (&gt;= 1).
     * The default value is {@code 4}.
     * </p>
     *
     * @see #rowsProperty()
     * @return the number of columns
     */
    public final IntegerProperty colsProperty() {
        return cols;
    }

    public final void setCols(int cols) {
        this.cols.set(cols);
    }

    private final StyleableIntegerProperty rows = new StyleableIntegerProperty(5) {
        @Override
        protected void invalidated() {
            if (get() < 1) {
                throw new IllegalArgumentException("number of rows must be larger than 0");
            }
        }
        @Override
        public Object getBean() { return YearView.this; }
        @Override
        public String getName() { return "rows"; }
        @Override
        public CssMetaData<? extends Styleable, Number> getCssMetaData() {
            return StyleableProperties.ROWS;
        }
    };

    public final int getRows() {
        return rows.get();
    }

    /**
     * Determines how many rows of years will be displayed. The number of
     * columns multiplied with the number of rows determines the total number
     * of years shown per "page".
     * <p>
     * Can be set via CSS using the {@code -fx-rows} property.
     * Valid values are positive integers (&gt;= 1).
     * The default value is {@code 5}.
     * </p>
     *
     * @see #colsProperty()
     * @return the number of rows
     */
    public final IntegerProperty rowsProperty() {
        return rows;
    }

    public final void setRows(int rows) {
        this.rows.set(rows);
    }

    private static class StyleableProperties {

        private static final CssMetaData<YearView, Number> COLS =
            new CssMetaData<>("-fx-cols", SizeConverter.getInstance(), 4) {
                @Override
                public boolean isSettable(YearView c) {
                    return !c.cols.isBound();
                }
                @Override
                public StyleableProperty<Number> getStyleableProperty(YearView c) {
                    return (StyleableProperty<Number>) c.cols;
                }
            };

        private static final CssMetaData<YearView, Number> ROWS =
            new CssMetaData<>("-fx-rows", SizeConverter.getInstance(), 5) {
                @Override
                public boolean isSettable(YearView c) {
                    return !c.rows.isBound();
                }
                @Override
                public StyleableProperty<Number> getStyleableProperty(YearView c) {
                    return (StyleableProperty<Number>) c.rows;
                }
            };

        private static final List<CssMetaData<? extends Styleable, ?>> STYLEABLES;
        static {
            List<CssMetaData<? extends Styleable, ?>> styleables = new ArrayList<>(Control.getClassCssMetaData());
            styleables.add(COLS);
            styleables.add(ROWS);
            STYLEABLES = Collections.unmodifiableList(styleables);
        }
    }

    public static List<CssMetaData<? extends Styleable, ?>> getClassCssMetaData() {
        return StyleableProperties.STYLEABLES;
    }

    @Override
    public List<CssMetaData<? extends Styleable, ?>> getControlCssMetaData() {
        return getClassCssMetaData();
    }

    private final ObjectProperty<Year> earliestYear = new SimpleObjectProperty<>(this, "earliestYear");

    public final Year getEarliestYear() {
        return earliestYear.get();
    }

    /**
     * The earliest year that the user will be able to select in the view.
     *
     * @return the earliest year
     */
    public final ObjectProperty<Year> earliestYearProperty() {
        return earliestYear;
    }

    public final void setEarliestYear(Year earliestYear) {
        this.earliestYear.set(earliestYear);
    }

    private final ObjectProperty<Year> latestYear = new SimpleObjectProperty<>(this, "latestYear");

    public final Year getLatestYear() {
        return latestYear.get();
    }

    /**
     * The latest year that the user will be able to select in the view.
     *
     * @return the earliest year
     */
    public final ObjectProperty<Year> latestYearProperty() {
        return latestYear;
    }

    public final void setLatestYear(Year latestYear) {
        this.latestYear.set(latestYear);
    }
}

File: docs/manuals/src/manuals/year_view.py
"""Content of the YearView developer manual."""

from manualkit import (
    Bullets, Callout, Chapter, Code, Figure, Manual, Numbered, Para,
    Property, PropertyTable, Section, Table,
)

G = "year-view"

MANUAL = Manual(
    control="YearView",
    package="com.dlsc.gemsfx",
    subtitle="A paged grid for selecting java.time.Year values",
    abstract=("YearView displays a configurable grid of years, marks the current and selected year, and pages through year ranges with arrow buttons."),
    cover_svg=f"{G}/cover.svg",
    cover_caption="Generated cartoon overview of YearView.",
    chapters=[
        Chapter("Introduction", [
            Para("<b>YearView</b> is part of the GemsFX module <font face='Courier'>com.dlsc.gemsfx</font>. This manual documents only behavior verified in the control, skin, CSS, resource bundles and demo app."),
            Section("Key features"),
            Bullets(['Selects a java.time.Year value, initialized to Year.now().', 'Read-only year property exposes the selected year as an int or -1 when value is null.', 'Rows and columns define how many years are shown per page.', 'Optional earliestYear and latestYear disable out-of-range years.', 'The skin shows a range label such as 2020-2039.']),
            Section("Maven dependency"),
            Code("""<dependency>
    <groupId>com.dlsc.gemsfx</groupId>
    <artifactId>gemsfx</artifactId>
    <version>4.4.1</version>
</dependency>""", caption="Use package <font face='Courier'>com.dlsc.gemsfx</font>."),
        ]),
        Chapter("Getting started", [
            Para("Create the control, set the properties that define its valid values, and listen to the value or selection property that the control owns."),
            Code("""YearView view = new YearView();
view.setCols(4);
view.setRows(5);
view.setEarliestYear(Year.of(2000));
view.setLatestYear(Year.now().plusYears(5));

view.valueProperty().addListener((obs, oldYear, newYear) -> {
    System.out.println("selected year = " + newYear);
});""", caption="A compact setup for <font face='Courier'>YearView</font>."),
            Figure(f"{G}/cover.svg", "A first look at the control in a simple application window."),
        ]),
        Chapter("Anatomy", [
            Para("The skin builds the visible nodes below the public control. The table lists the style classes and nodes that are useful when reading the source or writing CSS."),
            Figure(f"{G}/anatomy.svg", "The main parts of the control."),
            Table(["Part", "Style / node", "Description"], [
                        ["Root", "year-view", "Control root and stylesheet hook."],
                        ["Header", "header", "Left button, year-range label and right button."],
                        ["Range label", "year-range-label", "Shows first and last visible year on the current page."],
                        ["Grid", "grid-pane", "GridPane with rows x cols equally sized cells."],
                        ["Year box", "year-box", "Selectable VBox for one year."],
                        ["Selection indicator", "selection-indicator", "Small line shown when the box matches value/year."]
            ], widths=[18, 28, 54]),
        ]),
        Chapter("Control API", [
            Section("Selection"),
            PropertyTable([
                        Property("value", "ObjectProperty&lt;Year&gt;", "Year.now()", "Selected year."),
                        Property("year", "ReadOnlyIntegerProperty", "current year", "Integer value derived from value, or -1 if value is null."),
                        Property("earliestYear", "ObjectProperty&lt;Year&gt;", "null", "Earliest selectable year; earlier cells are disabled."),
                        Property("latestYear", "ObjectProperty&lt;Year&gt;", "null", "Latest selectable year; later cells are disabled.")
            ]),
            Section("Grid"),
            PropertyTable([
                        Property("cols", "IntegerProperty", "4", "Number of year columns. Must be at least 1. Styleable via -fx-cols."),
                        Property("rows", "IntegerProperty", "5", "Number of year rows. Must be at least 1. Styleable via -fx-rows.")
            ]),
        ]),
        Chapter("Behaviour", [
            Section("Paged ranges"),
            Para("The first visible year is calculated from the selected year or the current year, the visible cell count, and an internal page offset. Clicking arrows increments or decrements the offset."),
            Figure(f"{G}/behaviour.svg", "Interaction and value-flow behaviour."),
            Section("Range constraints"),
            Para("Year boxes bind their disabled state to earliestYear and latestYear. Clicking a disabled box is prevented by JavaFX disabled-node behavior."),
        ]),
        Chapter("Styling", [
            Para("The stylesheet is loaded by the control or its base class. The rows below list style hooks verified in the CSS and skin source."),
            Figure(f"{G}/styling.svg", "Style classes and pseudo-class states."),
            Section("Style classes"),
            Table(["Style class", "Where used"], [
                        ["year-view", "Root control style class."],
                        ["popover", "Optional root class for popup-specific padding and border."],
                        ["header", "Navigation row."],
                        ["year-range-label", "Header label showing the visible range."],
                        ["arrow-button left-button right-button", "Navigation buttons."],
                        ["arrow left-arrow right-arrow", "Arrow glyph regions."],
                        ["grid-pane", "GridPane containing year boxes."],
                        ["year-box", "Selectable year cell."],
                        ["year-label", "Label for the year number."],
                        ["selection-indicator", "Indicator below selected year."]
            ], widths=[35,65]),
            Section("Pseudo classes"),
            Table(["Pseudo class / marker", "Meaning"], [
                        ["focused", "Border switches to -fx-accent."],
                        ["selected", "Style class, not pseudo class, added to selected year-box."],
                        ["current", "Style class, not pseudo class, added to current year-box."]
            ], widths=[30,70]),
            Section("Styleable CSS properties"),
            Table(["Property", "Type", "Default", "Description"], [
                        ["-fx-cols", "Integer", "4", "Number of columns; positive integers only."],
                        ["-fx-rows", "Integer", "5", "Number of rows; positive integers only."]
            ], widths=[28,18,20,34]),
            Code(""".year-view {
    -fx-cols: 3;
    -fx-rows: 4;
}
.year-view > .grid-pane > .year-box > .selection-indicator {
    -fx-background-color: -fx-accent;
}""", caption="Example CSS."),
        ]),
        Chapter("Localization", [
            Para("This control uses <font face='Courier'>ResourceBundleManager</font> for the following keys."),
            Table(["Key", "English default"], [
                        ["accessible.role-description", "year selector"]
            ], widths=[55,45]),
        ]),
        Chapter("Accessibility", [
            Para("The constructor sets AccessibleRole.DATE_PICKER through AccessibilityUtil.setRole and uses the localized role description \"year selector\"."),
        ]),
        Chapter("Recipes", [
            Figure(f"{G}/recipes.svg", "Common configuration recipes."),
            Section("Show twelve years per page"),
            Code("""YearView view = new YearView();
view.setCols(3);
view.setRows(4);"""),
            Section("Restrict to a product lifetime"),
            Code("""view.setEarliestYear(Year.of(2020));
view.setLatestYear(Year.of(2030));"""),
            Section("Mirror the integer year"),
            Code("""Label label = new Label();
label.textProperty().bind(Bindings.convert(view.yearProperty()));"""),
            Section("Clear the selected value"),
            Code("""view.setValue(null); // yearProperty becomes -1"""),
            Section("Checklist"),
            Numbered(['Keep cols and rows >= 1.', 'Use earliestYear/latestYear for selectable limits.', 'Remember that selected/current are style classes on cells.', 'Use the value property for Year objects and yearProperty for int display.']),
        ]),
        Chapter("See also", [
            Para("Demo app: <font face='Courier'>YearViewApp</font>. Run it with:"),
            Code("mvn javafx:run -f gemsfx-demo/pom.xml -Dmain.class=com.dlsc.gemsfx.demo.YearViewApp"),
            Bullets([
                "Related GemsFX controls: YearPicker, YearMonthView, YearMonthPicker.",
                "API documentation: https://dlsc-software-consulting-gmbh.github.io/GemsFX/api/",
            ]),
        ]),
    ],
)
